src/control_server.rs

```diff
@@ -0,0 +1,75 @@
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use axum::{
+    extract::Json,
+    routing::{get, post},
+    Router,
+};
+use futures::FutureExt;
+use hyper::server::{self, conn::Http};
+use stream_cancel::Tripwire;
+use tokio::{net::UnixListener, sync::Mutex};
+use tokio_stream::wrappers::UnixListenerStream;
+
+use crate::{config::ConfigBuilder, topology::TopologyController};
+
+pub struct ControlServer {
+    listener: UnixListener,
+    topology_controller: Arc<Mutex<TopologyController>>,
+    shutdown_signal: Tripwire,
+    socket_path: PathBuf,
+}
+
+impl ControlServer {
+    pub fn bind(
+        socket_path: impl AsRef<Path>,
+        topology_controller: Arc<Mutex<TopologyController>>,
+        shutdown_signal: Tripwire,
+    ) -> Result<Self, crate::Error> {
+        // Try to remove any lingering socket from previous runs
+        if socket_path.as_ref().try_exists()? {
+            std::fs::remove_file(&socket_path)?;
+        }
+        let listener = UnixListener::bind(&socket_path)?;
+
+        Ok(ControlServer {
+            listener,
+            topology_controller,
+            shutdown_signal,
+            socket_path: socket_path.as_ref().to_path_buf(),
+        })
+    }
+
+    pub async fn run(self) -> Result<(), crate::Error> {
+        let app = Router::new()
+            .route("/ping", get(|| async { "pong" }))
+            .route(
+                "/config",
+                post(|Json(builder): Json<ConfigBuilder>| async move {
+                    if let Ok(mut controller) = self.topology_controller.try_lock() {
+                        let new_config = builder.build().ok();
+                        controller.reload(new_config).await;
+                        Ok(())
```

Comment:
Per Bruce's point, does this always returns a successful response regardless of the configuration validity or the reload success? Since we plan for OPW to incorporate the API in initial config validation (i.e. load the enterprise pipeline then "reload" the user config to start) and to re-report on successful reloads, could we return the errors?

Reply:
To expand on this a little, it would return errors for things like malformed or otherwise invalid configurations. That was pretty implicit in the axum handler argument types, which was not ideal. I've updated this now to make it more explicit, handle more types of errors, and make sure that we always return a JSON error response that we control and can evolve.

Reply:
Ah gotcha, right, the parameter was `ConfigBuilder`. Nice, thanks for updating this.